Feature/recycle

Nibuja05 opened the pull request:
Adds the recycling mechanic to the Anvil:

- Changed worth calculations of weapons to now be based on XP and represent shards instead of gold
- Added new random element, for various scenarios
- Added recycling option to anvil
- Added confirm mechanic anvil (used for recycling)
- Added options to show preview shards in the anvil output

Other Additions:
- Added round pause and the option to toggle it, by clicking on the timer

Other Fixes:
- Fixes #86 
- Fixed broken shard counter after absorbing

Shushishtok commented on content/panorama/scripts/custom_game/shop.ts:
```diff
@@ -636,21 +619,62 @@ function AnvilWeaponRemove(oldPanel: Panel) {
+	anvilCost.RemoveClass("Shown");
+
+	if (smooth) {
+		// do something
```
Do something, make it smooth :D
(seriously though, do something about that xD)

Nibuja05 replied:
ok, I realized that always smoothes out here, so no need to specify that ^^

Shushishtok commented on game/scripts/vscripts/core/shop_manager.ts:
```diff
@@ -330,4 +332,17 @@ export namespace ShopManager {
+		if (!weapon) return;
+
+		let shardCount = weapon.wGetWorth();
+		for (let i = 0; i <= shardCount; i++) {
```
Can this be changed in a way that gets remaining shards, rolls a number of shards (up to the max) and an element, then adds that? Less loops and events.

For instance, if I have 10 shards:
* Roll a number between 1 and 10 - say I get 3
* Roll an element - say I get Frost
* The player gets 3 Frost Shards, now there are 7 shards remaining
* Roll a number between 1 and 7 - say I get 5
* Roll an element - say I got Fire
* The player gets 5 Fire Shards, now there are 2 shards remaining
* Roll a number between 1 and 2 - say I get 2
* Roll an element - say I got Wind
* The player gets 2 Wind Shards, now there are 0 shards remaining and we can move on.

The process can be refined even further, e.g. max random shards per cycle can only be up to 5 in case you have more than 5, to make sure it doesn't just give all shards for a single element if you're lucky.

Nibuja05 replied:
used here a `DefaultMap` instead, works 👍

Shushishtok commented on game/scripts/vscripts/core/weapon_base.ts:
```diff
@@ -192,12 +192,9 @@ export class BaseWeapon {
-		}
-		return worth;
+		let baseWorth = 4; // needs to be adjusted later, if there are multiple rarities
+		let xpWorth = Math.floor(this.xp / 30);
```
Should be changed to something easily adjustable, e.g. `Math.floor(this.xp * 20 * 0.01 / 10)`.
The "20" above should be its own variable.

Nibuja05 commented:
- Added `Constants.ts` to keep track of all constant value we need to keep track of.
I changed all I found, but I might have missed some. We can continue to move them to there with time.

- Also added `DefaultMap` (inspired by pythons `DefautlDict`). Its the same as a normal map, improves the get function. More info at its documentation in `Util`.
Feel free to use it anywhere you want.

Shushishtok reviewed (changes requested):
Attempting to fix issue with pushing
